### PR TITLE
[BEAM-9542] Limit and clarify the effect of "force" in Java build

### PR DESCRIPTION
**Pull request by suztomo**
Before: forcing all dependencies for all class paths

After: forcing only 4 dependencies needed for tests.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by aaltay** [comment]
Run Java PostCommit

**Comment by aaltay** [comment]
Run Java HadoopFormatIO Performance Test

**Comment by aaltay** [comment]
Run BigQueryIO Streaming Performance Test Java

**Comment by aaltay** [comment]
Run Dataflow ValidatesRunner

**Comment by aaltay** [comment]
Run Spark ValidatesRunner

**Comment by aaltay** [comment]
Run SQL Postcommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -1258,23 +1258,20 @@ class BeamModulePlugin implements Plugin<Project> {
...
       // which chooses the latest version available.
       //
       // TODO: Figure out whether we should force all dependency conflict resolution
-      // to occur in the "shadow" and "shadowTest" configurations.
+      // to occur in the "shadowTest" configurations.
       project.configurations.all { config ->
-        // When running beam_Dependency_Check, resolutionStrategy should not be used; otherwise
-        // gradle-versions-plugin does not report the latest versions of the dependencies.
-        def startTasks = project.gradle.startParameter.taskNames
-        def inDependencyUpdates = 'dependencyUpdates' in startTasks || 'runBeamDependencyCheck' in startTasks
```
Comment (suztomo) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: I confirmed `./gradlew runBeamDependencyCheck` works as expected with this PR.
https://gist.github.com/suztomo/7077fcea856fba6761f17da3977b6c82

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 2]
```diff
@@ -1258,23 +1258,20 @@ class BeamModulePlugin implements Plugin<Project> {
...
-
-        // The "errorprone" configuration controls the classpath used by errorprone static analysis, which
-        // has different dependencies than our project.
-        if (config.getName() != "errorprone" && !inDependencyUpdates) {
```
Comment (suztomo) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Before this PR, the effect of `force` was unclear; this file had to special-case "errorprone" and "runBeamDependencyCheck" when unexpected behavior happened.

File: sdks/java/core/build.gradle [hunk 3]
```diff
@@ -87,6 +87,7 @@ dependencies {
   provided 'io.airlift:aircompressor:0.16' 
   provided 'com.facebook.presto.hadoop:hadoop-apache2:3.2.0-1'
   shadowTest library.java.jackson_dataformat_yaml
+  shadowTest library.java.junit
```
Comment (suztomo) [inline sdks/java/core/build.gradle]: This module uses JUnit.

File: sdks/java/io/google-cloud-platform/build.gradle [hunk 4]
```diff
@@ -75,6 +75,7 @@ dependencies {
   compile library.java.proto_google_common_protos
   compile library.java.protobuf_java
   compile library.java.slf4j_api
+  compileOnly library.java.hamcrest_library // should not appear in users' class path
```
Comment (suztomo) [inline sdks/java/io/google-cloud-platform/build.gradle]: This module's `org.apache.beam.sdk.io.gcp.pubsub.TestPubsub` in "main" (not test) uses Hamcrest's matchers. Therefore declaring the dependency.

Comment (lukecwik) [reply sdks/java/io/google-cloud-platform/build.gradle]: provided would make more sense

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 5]
```diff
@@ -1258,23 +1258,20 @@ class BeamModulePlugin implements Plugin<Project> {
...
+        String configName = config.getName()
+        if (configName.toLowerCase().contains('test')) {
           config.resolutionStrategy {
-            force project.library.java.values()
+            force project.library.java.junit,
+                    project.library.java.hamcrest_core,
+                    project.library.java.guava, // sdks:java:io:google-cloud-platform needs this
```
Comment (lukecwik) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: If its only needed in these projects, does it make sense to apply force rule there?

File: sdks/java/io/google-cloud-platform/build.gradle (hunk 4 above)
Comment (suztomo) [reply sdks/java/io/google-cloud-platform/build.gradle]: Updated to `provided`.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 5 above)
Comment (suztomo) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: I find centrally managing `force` more clear.

**Comment by je-ik** [comment]
Run Java PostCommit

**Comment by je-ik** [comment]
Run Java HadoopFormatIO Performance Test

**Comment by je-ik** [comment]
Run BigQueryIO Streaming Performance Test Java

**Comment by je-ik** [comment]
Run Dataflow ValidatesRunner

**Comment by je-ik** [comment]
Run Spark ValidatesRunner

**Comment by je-ik** [comment]
Run SQL Postcommit

**Comment by suztomo** [comment]
@je-ik Thank you. It seems Jenkins didn't trigger jobs and "BigQueryIO Streaming Performance Test Java 10 GB" still has problem. https://builds.apache.org/job/beam_BiqQueryIO_Streaming_Performance_Test_Java/664/

I'll check later.

**Comment by je-ik** [comment]
@suztomo do you need to rerun the checks that were not triggered?

**Comment by suztomo** [comment]
Yes, please

**Comment by je-ik** [comment]
Run Java PostCommit

**Comment by je-ik** [comment]
Run Java HadoopFormatIO Performance Test

**Comment by je-ik** [comment]
Run Dataflow ValidatesRunner

**Comment by je-ik** [comment]
Run Spark ValidatesRunner

**Comment by je-ik** [comment]
Run SQL Postcommit

**Comment by suztomo** [comment]
Because the effect of `force` is only for tests. The compileJava configuration still fetches the latest google-cloud-api, which is 1.30.8, which has a bug of android annotation dependency.

```
17:21:41 Execution failed for task ':sdks:java:testing:test-utils:compileJava'.
17:21:41 > Could not resolve all files for configuration ':sdks:java:testing:test-utils:compileClasspath'.
17:21:41    > Could not find androidx.annotation:annotation:1.1.0.
17:21:41      Searched in the following locations:
17:21:41        - file:/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Java_PR/src/sdks/java/testing/test-utils/offline-repository/androidx/annotation/annotation/1.1.0/annotation-1.1.0.pom
17:21:41        - file:/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Java_PR/src/sdks/java/testing/test-utils/offline-repository/androidx/annotation/annotation/1.1.0/annotation-1.1.0.jar
17:21:41        - https://repo.maven.apache.org/maven2/androidx/annotation/annotation/1.1.0/annotation-1.1.0.pom
17:21:41        - https://repo.maven.apache.org/maven2/androidx/annotation/annotation/1.1.0/annotation-1.1.0.jar
17:21:41        - file:/home/jenkins/.m2/repository/androidx/annotation/annotation/1.1.0/annotation-1.1.0.pom
17:21:41        - file:/home/jenkins/.m2/repository/androidx/annotation/annotation/1.1.0/annotation-1.1.0.jar
17:21:41        - https://jcenter.bintray.com/androidx/annotation/annotation/1.1.0/annotation-1.1.0.pom
17:21:41        - https://jcenter.bintray.com/androidx/annotation/annotation/1.1.0/annotation-1.1.0.jar
17:21:41        - https://oss.sonatype.org/content/repositories/staging/androidx/annotation/annotation/1.1.0/annotation-1.1.0.pom
17:21:41        - https://oss.sonatype.org/content/repositories/staging/androidx/annotation/annotation/1.1.0/annotation-1.1.0.jar
17:21:41        - https://repository.apache.org/snapshots/androidx/annotation/annotation/1.1.0/annotation-1.1.0.pom
17:21:41        - https://repository.apache.org/snapshots/androidx/annotation/annotation/1.1.0/annotation-1.1.0.jar
17:21:41        - https://repository.apache.org/content/repositories/releases/androidx/annotation/annotation/1.1.0/annotation-1.1.0.pom
17:21:41        - https://repository.apache.org/content/repositories/releases/androidx/annotation/annotation/1.1.0/annotation-1.1.0.jar
17:21:41      Required by:
17:21:41          project :sdks:java:testing:test-utils > project :sdks:java:extensions:google-cloud-platform-core > com.google.api-client:google-api-client:1.30.8
```

**Comment by suztomo** [comment]
Google-api-client version should be upgraded before this.

**Pull request closed without merging**